File: genesis/src/main/kotlin/com/omar/genesis/semester/service/SemesterService.kt
package com.omar.genesis.semester.service

import com.omar.genesis.semester.repository.SemesterRepository
import org.springframework.stereotype.Service

@Service
class SemesterService(private val repository: SemesterRepository) {
    fun getSemestersByCode(code: String) = repository.getSemestersByCode(code)
    fun getActiveSemesterByCode(code: String) = repository.getActiveSemester(code)
    fun getNonActiveSemesterByCode(code: String) = repository.getNonActiveSemester(code)
    fun paySemester(id: Long) = repository.paySemester(id)
}

File: genesis-frontend1/src/app/core/services/semester/semester.service.ts
import { Injectable } from '@angular/core';
import { AuthService } from '../auth/auth.service';
import { HttpClient } from '@angular/common/http';
import { Router } from '@angular/router';

@Injectable({
  providedIn: 'root',
})
export class SemesterService {
  private semester = 'http://localhost:8080/semester';

  constructor(
    private authService: AuthService,
    private http: HttpClient,
    private router: Router,
  ) {}

  getSemesters() {
    return this.http.get(`${this.semester}/code`, {
      params: { code: this.authService.account.code },
    });
  }

  getActiveSemester() {
    return this.http.get(`${this.semester}/active`, {
      params: { code: this.authService.account.code },
    });
  }

  getNonActiveSemester() {
    return this.http.get(`${this.semester}/non_active`, {
      params: { code: this.authService.account.code },
    });
  }

  paySemester(semester: any) {
    return this.http.get(`${this.semester}/pay`, {
      params: { id: semester.id },
    });
  }

  addAvailableCredits(semesterId: number, credits: number) {
    return this.http.get(`${this.semester}/add_available_credits`, {
      params: { id: semesterId, credits: credits },
    });
  }

  removeAvailableCredits(semesterId: number, credits: number) {
    return this.http.get(`${this.semester}/remove_available_credits`, {
      params: { id: semesterId, credits: credits },
    });
  }
}
